### PR TITLE
Improve targets

### PR DESCRIPTION
**Pull request by vadymmarkov**
Issue: https://github.com/hyperoslo/Form/issues/273

1) Now it's possible to change the following with targets:
- title
- info
- minimumDate
- maximumDate
- validation
- formula

2) `target_value` was changed to `value` in JSON (this is backwards compatible, we are checking for both for now)

### CODE REVIEW

File: Source/Models/FORMTarget.m [hunk 1]
```diff
@@ -11,18 +11,47 @@ @implementation FORMTarget
...
 - (instancetype)initWithDictionary:(NSDictionary *)dictionary {
-    self = [super init];
+    self = [super initWithDictionary:dictionary];
     if (!self) return nil;
 
     _targetID = [dictionary andy_valueForKey:@"id"];
     self.typeString = [dictionary andy_valueForKey:@"type"];
     self.actionTypeString = [dictionary andy_valueForKey:@"action"];
-    self.targetValue = [dictionary andy_valueForKey:@"target_value"];
+    self.value = [dictionary andy_valueForKey:@"value"];
```
Comment (3lvis) [inline Source/Models/FORMTarget.m]: Can we add backwards compatibility? Might be nice if we don't have to jump to Form 4.0.0

File: Source/Models/FORMTarget.h [hunk 2]
```diff
@@ -18,20 +19,22 @@ typedef NS_ENUM(NSInteger, FORMTargetActionType) {
...
 
 @property (nonatomic, copy) NSString *targetID;
-@property (nonatomic, copy) id targetValue;
+@property (nonatomic, copy) id value;
 @property (nonatomic, copy) NSString *typeString;
 @property (nonatomic, copy) NSString *actionTypeString;
 @property (nonatomic, copy) NSString *condition;
 
-@property (nonatomic) FORMFieldValue *value;
+@property (nonatomic) FORMFieldValue *fieldValue;
```
Comment (3lvis) [inline Source/Models/FORMTarget.h]: Can we add backwards compatibility? Might be nice if we don't have to jump to Form 4.0.0

You can just make `fieldValue` point to `value` with the deprecated flag.

```objc
- (NSInteger)foo:(NSInteger)bar __attribute__((deprecated("Use fooWithBar: instead")));
```

File: Source/Models/FORMTarget.m [hunk 3]
```diff
@@ -11,18 +11,47 @@ @implementation FORMTarget
...
+            values[@"value"] = self.value;
+        }
+
+        unsigned int numberOfProperties = 0;
+        objc_property_t *propertyArray = class_copyPropertyList([FORMFieldElement class], &numberOfProperties);
+
+        for (NSUInteger i = 0; i < numberOfProperties; i++) {
+            objc_property_t property = propertyArray[i];
+            NSString *propertyName = [[NSString alloc] initWithUTF8String:property_getName(property)];
+            NSLog(@"Prop: %@", propertyName);
```
Comment (3lvis) [inline Source/Models/FORMTarget.m]: Could you remove the NSLog?

File: Tests/Tests/FORMTargetTests.m [hunk 4]
```diff
@@ -85,4 +85,41 @@ - (void)testFilteredTargets {
...
+    NSDictionary *targetJSON = @{
+                                 @"id": @"end_date",
+                                 @"type": @"field",
+                                 @"action": @"update",
+                                 @"validations": @{ @"required": @NO },
+                                 @"condition": @"$employment_type == 2 && !missing($temporary_employee_reason)"};
+
+    FORMTarget *target = [[FORMTarget alloc] initWithDictionary:targetJSON];
+    NSDictionary *fieldProperties = [target fieldPropertiesToUpdate];
+    NSLog(@"Props: %@", fieldProperties);
```
Comment (3lvis) [inline Tests/Tests/FORMTargetTests.m]: Could you remove the NSLog?

File: Source/Models/FORMTarget.h (hunk 2 above)
Comment (vadymmarkov) [inline Source/Models/FORMTarget.h]: @3lvis There were the following changes:

 `targetValue` -> `value`
 `value` -> `fieldValue`

So I can't make `fieldValue` point to `value`. Maybe we could just back to previous names not to break backwards compatibility?

Comment (3lvis) [inline Source/Models/FORMTarget.h]: Ohh, you're right, my bad :)

**Comment by 3lvis** [comment]
@vadymmarkov there's a merge conflict :)

**Comment by 3lvis** [comment]
@vadymmarkov `FORMFieldElement` looks like it's something that could be used by itself, not something that is meant to be used as a base for a `FORMField` or a `FORMTarget`. Have you considered renaming it to `FORMFieldBase`?

**Comment by vadymmarkov** [comment]
@3lvis Yes, probably `FORMFieldBase` is better. Was thinking for a while what is the best name for this and ended up with `FORMFieldElement` don't know why. Still don't like inheritance in this case, but for now I think it's better to use some base class than extend `FORMField`.

**Comment by vadymmarkov** [comment]
@3lvis Merged with master

**Comment by 3lvis** [comment]
@vadymmarkov Awesome! Trying something, PR coming soon.

**Comment by vadymmarkov** [comment]
@3lvis Do we device not to merge it for now?

**Comment by 3lvis** [comment]
@vadymmarkov Need to fix https://github.com/hyperoslo/Form/pull/382 first :disappointed:

File: Source/FORMData.m [hunk 5]
```diff
@@ -806,71 +791,71 @@ - (NSArray *)updateTargets:(NSArray *)targets {
...
-                field = foundField;
-                if (indexPath) {
-                    [updatedIndexPaths addObject:indexPath];
+        if (field) {
+            NSArray *properties = [target propertiesToUpdate];
+            for (NSString *propertyName in properties) {
+                id value = [target valueForKey:propertyName];
+
+                if (![propertyName isEqualToString:@"value"]) {
+                    [field setValue:value forKey:propertyName];
```
Comment (3lvis) [inline Source/FORMData.m]: This is where the magic of this PR happens, here the target properties get set to the field

**Comment by vadymmarkov** [comment]
@3lvis I like it.

**Comment by 3lvis** [comment]
@vadymmarkov Great! Merged then :D

**Pull request merged**